File: 프로그래머스/lv2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
# from collections import deque
from collections import deque
def solution(bridge_length, weight, truck_weights):
    answer = 0
    que = deque(truck_weights)
    bridge = deque([0]*bridge_length)
    
    sum = 0
    while bridge or que:
        sum -= bridge.popleft()
        # truck_weights는 없고, bridge 남은 경우
        # 무게 초과한 경우
        if que:
            if (sum + que[0]) > weight:
                bridge.append(0)
            # 아닌경우
            else:
                tmp = que.popleft()
                bridge.append(tmp)
                sum += tmp
        answer += 1
    return answer

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;
class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        Queue<Integer> bridge = new LinkedList<>();
        for(int i = 0; i < bridge_length; i++) {
            bridge.offer(0);
        }
        
        Queue<Integer> trucks = new LinkedList<>();
        for(int i:truck_weights) {
            trucks.offer(i);
        }
        
        int sum = 0;
        while(bridge.size()!=0){
            sum-=bridge.peek();
            bridge.poll();
            if (trucks.size()!=0 && (sum + trucks.peek())<=weight){
                sum += trucks.peek();
                bridge.offer(trucks.poll());
            } else if(trucks.size()!=0){
                bridge.offer(0);
            }
            // System.out.println(bridge);
            answer++;
        }
        return answer;
    }
}